module/VuFind/src/VuFind/ApiKey/ApiKeyService.php

```diff
@@ -0,0 +1,173 @@
+ */
+class ApiKeyService implements DbServiceAwareInterface
+{
+    use DbServiceAwareTrait;
+
+    /**
+     * Constructor.
+     *
+     * @param Config $config Main config
+     */
+    public function __construct(
+        protected Config $config
+    ) {
+    }
+
+    /**
+     * Generate a new api key token
+     *
+     * @param UserEntityInterface $user User to create salt for
+     *
+     * @return string
+     */
+    protected function createRandomToken(UserEntityInterface $user): string
+    {
+        $salt = $this->config->API_Keys->token_salt ?? null;
+        if (!$salt) {
+            throw new \Exception('APIKeyService: Salt missing');
+        }
+        $valuesForToken = [
+            $user->getEmailVerified()->format('Y-m-d'),
+            $user->getFirstname(),
+            $user->getLastname(),
+            (string)strtotime('now'),
+            $salt,
+        ];
+        return hash('sha256', implode('|', $valuesForToken));
+    }
+
+    /**
+     * Retrieve an Api Key for a user. Return associative array containing token, revoked or empty
+     * array if not found.
+     *
+     * @param UserEntityInterface $user User
+     *
+     * @return ?AccessTokenEntityInterface
+     */
+    public function getApiKeyForUser(UserEntityInterface $user): ?AccessTokenEntityInterface
+    {
+        return $this->getDbService(AccessTokenService::class)->getByIdAndType(
+            (string)$user->getId(),
+            AccessTokenService::TYPE_API_KEY,
+            false
+        );
+    }
+
+    /**
+     * Check if Api Key is valid to be used.
+     *
+     * @param string $token Token to search for.
+     *
+     * @return bool If exists and the access token has not been revoked.
+     */
+    public function isTokenValid(string $token): bool
+    {
+        $token = $this->getDbService(AccessTokenService::class)->getByDataAndType(
+            $token,
+            AccessTokenService::TYPE_API_KEY
+        );
+        return $token && !$token->isRevoked();
+    }
+
+    /**
+     * Validate user can use api keys. It is expected that the user has a verified email address.
+     *
+     * @param UserEntityInterface $user User
+     *
+     * @return bool
+     */
+    public function isUserValid(UserEntityInterface $user): bool
+    {
+        return $user->getEmailVerified() !== null;
+    }
+
+    /**
+     * Generate an Api Key for a user.
+     *
+     * @param UserEntityInterface $user User
+     *
+     * @return string|false API key token on success, false on failure
+     */
+    public function generateApiKeyForUser(UserEntityInterface $user): string|false
+    {
+        // Check if the user has an existing token and the token has not been revoked.
+        $token = $this->getDbService(AccessTokenService::class)->getByIdAndType(
+            (string)$user->getId(),
+            AccessTokenService::TYPE_API_KEY
+        );
+        if ($token?->isRevoked()) {
```

Comment:
`$token?->isRevoked()` suggests that you expect that `$token` might be null... but subsequent code assumes it is an object. Should this be something like:

```suggestion
        if (!$token || $token->isRevoked()) {
```